plugin/pkg/scheduler/schedulercache/cache.go

```
@@ -0,0 +1,232 @@
+// "stop" is the channel that signals stopping and we would close background goroutines.
+func New(ttl, period time.Duration, stop chan struct{}) Cache {
+	cache := newSchedulerCache(ttl, period, stop)
+	cache.run()
+	return cache
+}
+
+type schedulerCache struct {
+	stop   chan struct{}
+	ttl    time.Duration
+	period time.Duration
+
+	// This mutex guards all fields within this cache struct
+	mu          sync.Mutex
+	assumedPods map[string]assumedPod
+	podStates   map[string]podState
+	nodes       map[string]*NodeInfo
+}
+
+type podState int
+
+const (
+	podAsummed podState = iota + 1
+	podAdded
+)
+
+type assumedPod struct {
+	pod      *api.Pod
+	deadline time.Time
+}
+
+func newSchedulerCache(ttl, period time.Duration, stop chan struct{}) *schedulerCache {
+	return &schedulerCache{
+		ttl:    ttl,
+		period: period,
+		stop:   stop,
+
+		nodes:       make(map[string]*NodeInfo),
+		podStates:   make(map[string]podState),
+		assumedPods: make(map[string]assumedPod),
+	}
+}
+
+func (cache *schedulerCache) GetNodeNameToInfoMap() map[string]*NodeInfo {
+	nodeNameToInfo := make(map[string]*NodeInfo)
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+	for name, info := range cache.nodes {
+		nodeNameToInfo[name] = info.Clone()
+	}
+	return nodeNameToInfo
+}
+
+func (cache *schedulerCache) List(selector labels.Selector) ([]*api.Pod, error) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	var pods []*api.Pod
+	for _, info := range cache.nodes {
+		pods = append(pods, info.pods...)
+	}
+	return pods, nil
+}
+
+func (cache *schedulerCache) AssumePodIfBindSucceed(pod *api.Pod, bind func() bool) error {
+	return cache.assumePodIfBindSucceed(pod, bind, time.Now())
+}
+
+// assumePodScheduled exists for making test deterministic by taking time as input argument.
+func (cache *schedulerCache) assumePodIfBindSucceed(pod *api.Pod, bind func() bool, now time.Time) error {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	if !bind() {
+		return nil
+	}
+
+	key := mustGetPodKey(pod)
```

To return error is not intended, because this error case is not transient error (e.g. data corruption). The pod must have a unique key, e.g. a name. Pod.Name is unique within Pod.Namespace. If some changes breaks this compatibility, it is very dangerous and we shouldn't swallow the error. 

We can extract namespace and name from api.Pod directly. But depending on MetaNamespaceKeyFunc seems more maintainable.

Not quite sure about this part. Just let you know that we shouldn't let such error not handled instantaneously.